### PR TITLE
Modularization

### PR DESCRIPTION
**Pull request by FrankBuchholz**
Hi Rudy,
here is a first proposal for modularization - the Lua scripts of the track layout contain only the data, all code is moved to a module..

I tried to transform global variables into local variables and renamed many of them for better raeadibility (at least for me...).
Sometimes I added more local variables instead of direct addressing within one statement.

I tried to get rid of the MAX constant (i do not like if programs invent arbitrary limitations). I was sucessful by choosing a dummy train number -999 instead (but for unknown reasons I wasn't able to use a smaller number like -4 as 'magic' constant because then train will clash. I've no idea why this happens. The 0-index-is-trash-trick is quite smart but I've to confess that I do not fully understand all implications. 

I added some consistency checks and additional tipp texts on the signals (which you can toggle using the main switch) to better unterstand what happens. I wasn't able to solve the pause/reload Lua/save-reload-issue yet.

Unfortunately this means that I've touched a lot of stuff... 
What do you think about this version ()

Kind regards
Frank

### CODE REVIEW

**Comment by RudyB24** [comment]
Frank,

Sounds like nice work ... I'll look into it. Which may take some time ...
besides preparing video 5, I have a busy week ahead.

I don't know Github well enough ... is there a way for your code to be
available to anyone who likes to download it, and at the same time for the
current code to stay available? I like to anyhow keep the old code too, so
I still know where to find things when I like to make enhancements. :)

One of the challenges still on the todo list is to be able to define 'via'
blocks. This can be necessary on certain layouts where trains otherwise
could run in a deadlock, when opposing trains noth have nowhere to go to.
With 'via' blocks a train can look more than one block ahead and only only
flag the multo-block route as available when all blocks on the way, and
specifically the final destination block are not reserved. Maybe you have
ideas how to tackle this. If that is tackled, than really all layouts can
be automated. The rest is more or less 'programming cosmetics'.

Greetings, Rudy.

File: LUA/blockControl.lua [hunk 1]
```diff
@@ -29,17 +29,24 @@ twowayblk = { 0, 3, 2, 0, 0, 0} -- Blocks which are used in both directions. Ent
 blocksig  = { 8,18, 9,19,10,13} -- Block signals
 memsig    = { 5, 6, 7,15,20,21} -- Corresponding memory signal per block
 
--- Configure possible routes between blocks here
```
Comment (FrankBuchholz) [inline LUA/blockControl.lua]: The index numbers of table 'route' are not important. Therefore, I like to get rid of the requirement to define data having a full set of index numbers without holes.
Example:
T = {}
T[1] = "apples"
T[2] = "bananas"
T[10] = "peaches"
print( #T )    --> 2
print( T[1] )  --> apples
print( T[2] )  --> bananas
print( T[3] )  --> nil
print( T[10] ) --> peaches

for k=1, #T do print(k, " ",T[k]) end
-->1 apples, 2 bananas

for k, v in pairs(T) do print(k, " ",T[k]," ",v) end
-->1 apples, apples 2 bananas bananas, 10 peaches peaches

File: LUA/blockControl.lua [hunk 2]
```diff
@@ -322,7 +332,8 @@ local function run ()
   
   showSignalStatus()                    -- Show current signal status of all block signals
 
-  local fr = 0                          -- Initialization of count of free routes
+  local available = {} 					-- Stores available routes. Per EEPmain() cycle only one route will be randomly selected fom this table
```
Comment (FrankBuchholz) [inline LUA/blockControl.lua]: We do not need variable 'fr' which is identical to '#available' if we reset this table properly.

File: LUA/blockControl.lua [hunk 3]
```diff
@@ -12,7 +12,7 @@
 
 -- Table of named trains having an optional specific train signal to start/stop this train individually.
 -- (You can omit the "#" as first character.)
-train = {}
```
Comment (FrankBuchholz) [inline LUA/blockControl.lua]: I had used my default settings for indention (tabs with visual length 4) in contrast to your settings (2 spaces).
I now switched to your style and reformatted the code.
Unfortunately I had missed to create a separate commit for this change.

File: LUA/blockControl.lua [hunk 4]
```diff
@@ -31,23 +31,38 @@ memsig    = { 5, 6, 7,15,20,21} -- Corresponding memory signal per block
...
--- Using the notation above we define the whole table in one Lua statement (spanning from line 33 to 43). 
+
+-- Using the notation above we define the whole table in one Lua statement (spanning from line 33 to 43).
 -- Instead of this you can use multiple Lua statements  using indices as well:
 --route = {}
 --route[1] = { 1, 6, turn={  2,1, 12,1       }}
 --route[2] = { 2, 6, turn={  2,2, 12,1, 17,1 }}
 --...
 
+-- Sometimes a lockdown could occur if trains just travel from block to block.
```
Comment (FrankBuchholz) [inline LUA/blockControl.lua:52]: This is the important part of this commit about destination blocks.

File: LUA/blockControl.lua [hunk 5]
```diff
@@ -400,38 +413,57 @@ local function run ()
...
+              and freeRoute then                                     -- Check if the destination block is free
+
+                for to = 1, #route[r].turn / 2 do               -- Check if the route turnouts are free (free==0)
+                  local switch = route[r].turn[to*2-1]
+                  if turnReserved[ switch ] then                -- Alternative compact Lua code: freeRoute = freeRoute and turnReserved[ switch ]
+                    freeRoute = false 
+                  end
                 end
-              end
+
```
Comment (FrankBuchholz) [inline LUA/blockControl.lua:429]: This is the important part of this commit about destination blocks.

**Pull request closed without merging**